exporter/prometheus/prometheus.go

```diff
@@ -145,151 +115,163 @@ type collector struct {
-	c.registeredViewsMu.Lock()
-	registered := make(map[string]*prometheus.Desc)
-	for k, desc := range c.registeredViews {
-		registered[k] = desc
-	}
-	c.registeredViewsMu.Unlock()
-
-	for _, desc := range registered {
-		ch <- desc
-	}
+	de := &descExporter{c: c, descCh: ch}
+	c.reader.ReadAndExport(de)
 }
 
 // Collect fetches the statistics from OpenCensus
 // and delivers them as Prometheus Metrics.
-// Collect is invoked everytime a prometheus.Gatherer is run
+// Collect is invoked every time a prometheus.Gatherer is run
 // for example when the HTTP endpoint is invoked by Prometheus.
 func (c *collector) Collect(ch chan<- prometheus.Metric) {
-	// We need a copy of all the view data up until this point.
-	viewData := c.cloneViewData()
-
-	for _, vd := range viewData {
-		sig := viewSignature(c.opts.Namespace, vd.View)
-		c.registeredViewsMu.Lock()
-		desc := c.registeredViews[sig]
-		c.registeredViewsMu.Unlock()
-
-		for _, row := range vd.Rows {
-			metric, err := c.toMetric(desc, vd.View, row)
-			if err != nil {
-				c.opts.onError(err)
-			} else {
-				ch <- metric
-			}
-		}
-	}
-
+	me := &metricExporter{c: c, metricCh: ch}
+	c.reader.ReadAndExport(me)
 }
 
-func (c *collector) toMetric(desc *prometheus.Desc, v *view.View, row *view.Row) (prometheus.Metric, error) {
-	switch data := row.Data.(type) {
-	case *view.CountData:
-		return prometheus.NewConstMetric(desc, prometheus.CounterValue, float64(data.Value), tagValues(row.Tags, v.TagKeys)...)
-
-	case *view.DistributionData:
-		points := make(map[float64]uint64)
-		// Histograms are cumulative in Prometheus.
-		// Get cumulative bucket counts.
-		cumCount := uint64(0)
-		for i, b := range v.Aggregation.Buckets {
-			cumCount += uint64(data.CountPerBucket[i])
-			points[b] = cumCount
-		}
-		return prometheus.NewConstHistogram(desc, uint64(data.Count), data.Sum(), points, tagValues(row.Tags, v.TagKeys)...)
+func newCollector(opts Options, registrar *prometheus.Registry) *collector {
+	return &collector{
+		reg:    registrar,
+		opts:   opts,
+		reader: metricexport.NewReader()}
+}
 
-	case *view.SumData:
-		return prometheus.NewConstMetric(desc, prometheus.UntypedValue, data.Value, tagValues(row.Tags, v.TagKeys)...)
+func (c *collector) toDesc(metric *metricdata.Metric) *prometheus.Desc {
+	return prometheus.NewDesc(
+		metricName(c.opts.Namespace, metric),
+		metric.Descriptor.Description,
+		toPromLabels(metric.Descriptor.LabelKeys),
+		c.opts.ConstLabels)
+}
 
-	case *view.LastValueData:
-		return prometheus.NewConstMetric(desc, prometheus.GaugeValue, data.Value, tagValues(row.Tags, v.TagKeys)...)
+type metricExporter struct {
+	c        *collector
+	metricCh chan<- prometheus.Metric
+}
 
-	default:
-		return nil, fmt.Errorf("aggregation %T is not yet supported", v.Aggregation)
+// ExportMetrics exports to the Prometheus.
+// Each OpenCensus Metric will be converted to
+// corresponding Prometheus Metric:
+// SumData will be converted to Untyped Metric,
```

Comment:
These comments no longer apply here.

Reply:
fixed. I have referred to Metric Types (like TypeCumulativeInt64)  in modified comment instead of stats aggregation type (like sum).